devito/arguments.py

```diff
@@ -158,198 +107,350 @@ class PtrArgument(Argument):
-        end = ScalarArgument(self.end_name, self, max)
-        return namedtuple("RuntimeArguments", ["size", "start", "end"])(size, start, end)
-
-    # TODO: Can we do without a verify on a dimension?
-    def verify(self, value, enforce=False):
+    def __init__(self, stencils, parameters, dle_arguments):
+        self.stencils = stencils
+        self.parameters = parameters
+        self.dle_arguments = dle_arguments
+        self.argument_mapper = self._build_argument_mapper(parameters)
+        self.arguments = filter_ordered([x for x in self.argument_mapper if isinstance(x, Argument)], key=lambda x: x.name)
+        self.dims = [x for x in self.argument_mapper if isinstance(x, DimensionParameter)]
+        self.offsets = {d.end_name: v for d, v in retrieve_offsets(stencils).items()}
+
+    def handle(self, **kwargs):
+
+        user_autotune = kwargs.pop('autotune', False)
+
+        kwargs = self._offset_adjust(kwargs)
+        
+        kwargs = self._extract_children_of_composites(kwargs)
+
+        values = self._derive_values(kwargs)
+
+        # The following is only being done to update the autotune flag. The actual value derivation for the
+        # dle arguments has moved inside the above _derive_values method.
+        # TODO: Refactor so this is not required
+        dim_sizes = dict([(d.name, runtime_dim_extent(d, values)) for d in self.dimensions])
+        dle_arguments, dle_autotune = self._dle_arguments(dim_sizes)
+        
+
+        assert(self._verify(values))
+
+        arguments = OrderedDict([(k.name, v) for k, v in values.items()])
+        return arguments, user_autotune and dle_autotune
+
+    def _offset_adjust(self, kwargs):
+        for k, v in kwargs.items():
+            if k in self.offsets:
+                kwargs[k] = v + self.offsets[k]
+        return kwargs
+
+    def _build_argument_mapper(self, parameters):
+        # Pass through SymbolicFunction
+        symbolic_functions = [x for x in parameters if isinstance(x, SymbolicFunction)]
+        dimension_dependency_mapper = dict()
+        tensor_arguments = []
+        for f in symbolic_functions:
+            argument = ArgumentVisitor().visit(f)
+            tensor_arguments.append(argument)
+            for i, d in enumerate(f.indices):
+                if d not in dimension_dependency_mapper:
+                    dimension_dependency_mapper[d] = []
+                dimension_dependency_mapper[d].append(Dependency("gets_value_from", argument, param=i))
+
+        for arg in self.dle_arguments:
+            d = arg.argument
+            if d not in dimension_dependency_mapper:
+                    dimension_dependency_mapper[d] = []
+            dimension_dependency_mapper[d].append(Dependency("gets_value_from", derive_dle_argument_value, param=arg))
+            
+        # Record dependencies in Dimensions
+        dimension_parameter_mapper = {}
+        for dim, deps in dimension_dependency_mapper.items():
```

Comment:
not pythonic.
An alternative is:
``mapper = {k: DimensionParameter(k, v) for k, v in dimension_dependency_mapper.items()}``

Reply:
Fixed